### ISSUE TITLE
Add back in `lm_head.weight` in Qwen2 for tied models

### ISSUE THREAD

**Issue by joecummings** [pytorch/torchtune#1381]

**Comment by SalmanMohammadi** [comment]
For the 0.5B and 1.B models? For why?

**Comment by joecummings** [comment]
Sorry, I wrote this while messing around with loading into HF.

This specifically related to the conversion *back* to the HF formatted weights. See this line: https://github.com/pytorch/torchtune/blob/9e65fa9ae67ce09e953cb68b5093d42cb6b31310/torchtune/models/qwen2/_convert_weights.py#L82. (I edited the Issue title to be clearer)

So when we load the model, we skip loading in the output and embed both b/c they're tied. So, when we remap to HF weights in our conversion, we don't have an output key and therefore it's not mapped to a lm_head key. HOWEVER, the HF model expects an lm_head key even if it's tied. So we need to just modify the conversion script to add it back in after training but before saving the model.

Does this make sense?

**Comment by SalmanMohammadi** [comment]
How are you loading into HF here? load_state_dict or from_pretrained?

**Comment by SalmanMohammadi** [comment]
```python
from torchtune.models import qwen2 
from torchtune import utils
import torch
checkpointer = utils.FullModelHFCheckpointer(
    checkpoint_dir="/home/salman/models/Qwen2-1.5B-Instruct/",
    checkpoint_files=[
        "model.safetensors",
    ],
    model_type="QWEN2",
    output_dir="/tmp/model_tmp",
)
with utils.set_default_dtype(torch.bfloat16):
    model = qwen2.qwen2_1_5b()
state_dict = checkpointer.load_checkpoint()["model"]
model.load_state_dict(state_dict=state_dict)

checkpointer.save_checkpoint({"model": model.state_dict() }, epoch=10)

# Load model directly
from transformers import Qwen2ForCausalLM
# need to mv /tmp/model_tmp/hf_model_0001_10.pt /tmp/model_tmp/pytorch_model.bin
model = Qwen2ForCausalLM.from_pretrained("/tmp/model_tmp", torch_dtype=torch.bfloat16)
🤗🤗🤗🤗🤗
```
Like Gemma, Qwen2 1.5B and 0.5B have tied weights and the checkpoints don't actually contain lm_head https://huggingface.co/Qwen/Qwen2-1.5B/tree/main?show_file_info=model.safetensors. `from_pretrained` correctly handles this logic. re: https://github.com/pytorch/torchtune/issues/1122

lmk if I'm missing something here. while we're here, I'd be interested in helping out with streamlining torchtune -> transformers checkpointing if that's on the cards, it's a bit non-obvious imo (had to find https://github.com/pytorch/torchtune/issues/832 and https://github.com/pytorch/torchtune/issues/878)

**Comment by joecummings** [comment]
> ```python
> from torchtune.models import qwen2 
> from torchtune import utils
> import torch
> checkpointer = utils.FullModelHFCheckpointer(
>     checkpoint_dir="/home/salman/models/Qwen2-1.5B-Instruct/",
>     checkpoint_files=[
>         "model.safetensors",
>     ],
>     model_type="QWEN2",
>     output_dir="/tmp/model_tmp",
> )
> with utils.set_default_dtype(torch.bfloat16):
>     model = qwen2.qwen2_1_5b()
> state_dict = checkpointer.load_checkpoint()["model"]
> model.load_state_dict(state_dict=state_dict)
> 
> checkpointer.save_checkpoint({"model": model.state_dict() }, epoch=10)
> 
> # Load model directly
> from transformers import Qwen2ForCausalLM
> # need to mv /tmp/model_tmp/hf_model_0001_10.pt /tmp/model_tmp/pytorch_model.bin
> model = Qwen2ForCausalLM.from_pretrained("/tmp/model_tmp", torch_dtype=torch.bfloat16)
> 🤗🤗🤗🤗🤗
> ```
> 
> Like Gemma, Qwen2 1.5B and 0.5B have tied weights and the checkpoints don't actually contain lm_head [huggingface.co/Qwen/Qwen2-1.5B/tree/main?show_file_info=model.safetensors](https://huggingface.co/Qwen/Qwen2-1.5B/tree/main?show_file_info=model.safetensors). `from_pretrained` correctly handles this logic. re: #1122
> 
> lmk if I'm missing something here. while we're here, I'd be interested in helping out with streamlining torchtune -> transformers checkpointing if that's on the cards, it's a bit non-obvious imo (had to find #832 and #878)

You are absolutely correct, this is solved with `from_pretrained` - I was trying to do this via state dict mapping. God, we need to make this more clear. Closing for now. Thanks for your help!
